Add tests for alert and anomaly generators

melodychn opened the pull request:
Add tests for `SimpleAlertGenerator` and `SimpleAnomalyGenerator` that is in pull request (#19).

Tests added for all `public` methods besides constructors and `static` create methods which are not needed for now (after discussion with Paula). 

The expected anomaly/alert are hard-coded in order to check for logic correctness of the generators. A shortened version of the csv data in the form of a string is used as input to the `SimpleAlertGenerator` in order for tests to always run on same data and easier to manually find expected anomalies. 

Some map/multimap implementations are changed to `LinkedHashMap` and `LinkedListMultimap` in order to maintain chronological order for the `Alert`s. Chronological order based on `Timestamp` is preferred for the `Anomaly` List within an `Alert` and for `Alert`s in a list of `Alert`s, so they can be stored in order within datastore and eventually sent in order to frontend.

cyu6 commented on backend/src/main/java/com/google/blackswan/mock/SimpleAnomalyGenerator.java:
```diff
@@ -59,7 +59,14 @@ private SimpleAnomalyGenerator(InputStream source, int threshold,
         .filter(entry -> entry.getValue() - avg > threshold)
-        .collect(toMap(entry -> entry.getKey(), entry -> entry.getValue()));
+        .collect(toMap(
+            entry -> entry.getKey(), 
+            entry -> entry.getValue(),
+            // In case of conflicting keys, pick value of second key. 
+            // But there should not be conflicting keys at all. 
+            (x, y) -> y,
+            LinkedHashMap::new
+          ));
```
to confirm: this is still mapping key to value and the exact map is a `LinkedHashMap`, but the `(x, y) -> y` means to pick the value of the second key?

melodychn replied:
yep, that is correct

cyu6 commented on backend/src/test/java/com/google/blackswan/mock/SimpleAlertGeneratorTest.java:
```diff
@@ -0,0 +1,117 @@
+
+  private static String inputForAnomalyGenerator() {
+    StringBuilder str = new StringBuilder("");
+    for (Map.Entry<String, Integer> entry : SAMPLE_DATA.entrySet()) {
+      str.append(entry.getKey() + "," + entry.getValue().toString() + "\n");
+    }
```
this is the fake csv data, right?

melodychn replied:
yes, this is the testing csv data

paulaksp replied:
might be good to make that clear in the fn comment that the inputForAnomalyGenerator is mimic-ing the csv format.

cyu6 commented on backend/src/test/java/com/google/blackswan/mock/SimpleAlertGeneratorTest.java:
```diff
@@ -0,0 +1,117 @@
+      new Timestamp("2019-09-01"), 
+      Arrays.asList(expectedAnomalyGroup2_1, expectedAnomalyGroup2_2), 
+      Alert.StatusType.UNRESOLVED
+    );
```
jw did you basically just copy over the data the alert generator would use to generate the alerts?

melodychn replied:
yes, i copied a section of the data, so it is easier for me to test, instead of the whole file

this also makes the test more stable as it does not depend on external files

cyu6 commented on backend/src/test/java/com/google/blackswan/mock/SimpleAnomalyGeneratorTest.java:
```diff
@@ -0,0 +1,120 @@
+    anomalyGenerator = SimpleAnomalyGenerator.createGeneratorWithString(
+      inputForAnomalyGenerator(), SET_THRESHOLD_HIGH, SET_DATAPOINTS
+    );
+    // Avg = 70, only 1 data point exceed 76. 
```
vvvvv nit: exceeds

cyu6 commented on backend/src/test/java/com/google/blackswan/mock/SimpleAnomalyGeneratorTest.java:
```diff
@@ -0,0 +1,120 @@
+  }
+
+  @Test
+  public void getAnomalies_returnsListOfAnomaliesWithSizeTwo() {
```
is it important to have another test? seems like it's testing the same thing as the SizeOne test.

melodychn replied:
I feel like it is important to have this test, because there will be two anomalies returned instead of one. also, the data points map included with the two anomalies all have edge cases. i wanted to test in addition to two anomalies being returned, whether the data points map gets the correct points for these edge cases.

cyu6 replied:
ah okay. thanks for clarifying!

paulaksp commented on backend/src/test/java/com/google/blackswan/mock/SimpleAlertGeneratorTest.java:
```diff
@@ -0,0 +1,117 @@
+  );
+  private static final int SET_THRESHOLD_LOW = 2;
+  private static final int SET_DATAPOINTS = 2;
+  private static final int EXPECTED_ALERT_SIZE = 2;
```
v nit: newline between the static constants and non-static constants

paulaksp commented on backend/src/test/java/com/google/blackswan/mock/SimpleAlertGeneratorTest.java:
```diff
@@ -0,0 +1,117 @@
+
+    List<Alert> generatedAlerts = alertGenerator.getAlerts();
+
+    assertEquals(generatedAlerts.size(), EXPECTED_ALERT_SIZE);
+    assertEquals(generatedAlerts.get(0), expectedAlertJuly);
+    assertEquals(generatedAlerts.get(1), expectedAlertAugust);
```
assertEquals(Object expected, Object actual)
Asserts that two objects are equal.


and all the other non-object primitives, the assertEquals are in the format expected first and then actual later.

melodychn replied:
thanks for letting me know! I will fix this from now on.

paulaksp commented on backend/src/test/java/com/google/blackswan/mock/SimpleAnomalyGeneratorTest.java:
```diff
@@ -0,0 +1,120 @@
+  private static final int SET_THRESHOLD_HIGH = 6;
+  private static final int SET_THRESHOLD_LOW = 4;
+  private static final int SET_DATAPOINTS = 2;
+  private final LocalServiceTestHelper helper =
```
same here

paulaksp commented on backend/src/test/java/com/google/blackswan/mock/SimpleAnomalyGeneratorTest.java:
```diff
@@ -0,0 +1,120 @@
+    assertEquals(generatedAnomalies.get(1), expectedAnomaly2);
+  }
+
+  private static String inputForAnomalyGenerator() {
+    StringBuilder str = new StringBuilder("");
+    for (Map.Entry<String, Integer> entry : SAMPLE_DATA.entrySet()) {
+      str.append(entry.getKey() + "," + entry.getValue().toString() + "\n");
+    }
+    return str.toString();
+  }
```
Alternatively, since the exact code is being written for both tests, you can also have test helper classes/files which are specifically only for test purposes.

ie instead of private static String inputForAnomalyGenerator(), you can have the static String inputForAnomalyGenerator(map_structure), and reuse it in both tests.

melodychn replied:
yep, i was considering creating a helper class/file before, but I wasn't too sure if it was too much for just one function. 

but i think it is definitely good practice, since the two functions are exactly the same!

paulaksp commented on backend/src/test/java/com/google/blackswan/mock/SimpleAnomalyGeneratorTest.java:
```diff
@@ -0,0 +1,120 @@
+
+    List<Anomaly> generatedAnomalies = anomalyGenerator.getAnomalies();
+
+    assertEquals(generatedAnomalies.size(), 2);
+    assertEquals(generatedAnomalies.get(0), expectedAnomaly1);
+    assertEquals(generatedAnomalies.get(1), expectedAnomaly2);
```
same here with the expected/actual ordering.

paulaksp commented on backend/src/test/java/com/google/blackswan/mock/SimpleAnomalyGeneratorTest.java:
```diff
@@ -0,0 +1,120 @@
+import javax.servlet.http.HttpServletResponse;
+import com.google.appengine.tools.development.testing.LocalDatastoreServiceTestConfig;
+import com.google.appengine.tools.development.testing.LocalServiceTestHelper;
+
```
nit: probably don't need the newline between the import statements here.

melodychn replied:
yep, i got rid of it. thanks!

i didnt get rid of the new line in line 18 as i think it separates the imports that are used for testing versus other imports for data structures and classes that are being tested.

melodychn commented:
i realized that many of my previous tests had the order of the 
```
assertEquals(Object expected, Object actual)
```
flipped, so i also fixed the order for my previous tests! hopefully its not confusing that i fixed it within the same pr, because i think it is not very big of a change...
